File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses, 0);
        vector<vector<int>> graph(numCourses);
        for (auto &v : prerequisites) {
            indegree[v[0]]++;
            graph[v[1]].push_back(v[0]);
        }
        vector<int> res;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                res.push_back(i);
            }
        }
        for (int i = 0; i < (int) res.size(); i++) {
            int v = res[i];
            for (auto& adj : graph[v]) {
                indegree[adj]--;
                if (indegree[adj] == 0) {
                    res.push_back(adj);
                }
            }
        }
        return res.size() == numCourses;
    }
};

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> adj = new HashMap<>();
        for (int i = 0; i < numCourses; i++) {
            adj.put(i, new ArrayList<>());
        }
        int[] indegree = new int[numCourses];
        for (int[] pre : prerequisites) {
            adj.get(pre[1]).add(pre[0]);
            indegree[pre[0]] += 1;
        }
        Queue<Integer> queue = new LinkedList<>();
        int canBeCompletedCount = 0;
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0) {
                queue.add(i);
                canBeCompletedCount += 1;
            }
        }
        while (!queue.isEmpty()) {
            Integer node = queue.poll();
            for (Integer neigh : adj.get(node)) {
                indegree[neigh] -= 1;
                if (indegree[neigh] == 0) {
                    queue.add(neigh);
                    canBeCompletedCount += 1;
                }
            }
        }
        return canBeCompletedCount == numCourses;
    }
}